astlib/migrate_501_500.ml

```diff
@@ -220,12 +220,51 @@ and copy_value_binding :
  fun {
        Ast_501.Parsetree.pvb_pat;
        Ast_501.Parsetree.pvb_expr;
+       Ast_501.Parsetree.pvb_constraint;
        Ast_501.Parsetree.pvb_attributes;
        Ast_501.Parsetree.pvb_loc;
      } ->
+  let pvb_pat = copy_pattern pvb_pat and pvb_expr = copy_expression pvb_expr in
+  let pvb_pat, pvb_expr =
+    match pvb_constraint with
+    | None -> (pvb_pat, pvb_expr)
+    | Some { locally_abstract_univars; typ } ->
+        let typ = copy_core_type typ in
+        let typ_poly =
+          {
+            typ with
+            ptyp_attributes = [];
```

Comment:
Why do you remove the attributes here and everywhere else? Are they duplicated, and you're de-duplicating them?

Reply:
Cool, thanks for explaining!

About the locations: to make sure that the migrated parsetree fulfills the two location invariants  "non-ghost children are nested" and "non-ghost siblings are disjoint and ordered", you could run a trivial ppxlib driver with `-perform_locations_check`. With "trivial" driver I mean one with no PPXs linked. However, I think that those two invariants should be fulfilled here anyways. You'll know better than me! I don't think we have more location checks than that.

> If there is a convenient way to test this, and get a diff, that would be great!

You might already have thought about the following, but just in case: The following should work I think. You could write a few-liner program using Ppxlib, which parses a file and pprints its ast. Then, to get the "parsing with 5.1 then migrating to 5.0" AST, you can compile that with OCaml `trunk` and this Ppxlib branch. And for the "parsing with 5.0", you can compile it with OCaml 5.0 and the `main` Ppxlib branch. Btw, if we want to take more time at some point, it would also be valuable to make a CI job out of this!

(About the "parsing with 5.1 then migrating to 5.0" part: When pprinting an AST, Ppxlib pprints the AST in its Ppxlib version, i.e. currently 5.0/4.14. In my opinion, that's not an ideal choice of ours, but that's a different topic...Also, in this particular situation it comes in nice.)